pyfolio/utils.py

```diff
@@ -108,29 +95,31 @@ def default_returns_func(symbol):
         Daily returns for the symbol.
          - See full explanation in tears.create_full_tear_sheet (returns).
     """
+    start = '1/1/1970' if start is None else start
+    end = datetime.now() if end is None else end
+    start = pd.Timestamp(start)
+    end = pd.Timestamp(end)
 
-    rets = None
-    if symbol == 'SPY':
-        filepath = data_path('spy.h5')
-        try:
-            # If it's been less than a day since we got benchmark
-            if datetime.now() - pd.to_datetime(
-                    getmtime(filepath), unit='s') < pd.Timedelta(days=1):
-                rets = pd.read_hdf(filepath, 'df')
-        except:
-            pass
-
-    if rets is None:
-        px = web.get_data_yahoo(symbol, start='1/1/1970')
+    def get_symbol_from_yahoo(symbol, start=None, end=None):
+        px = web.get_data_yahoo(symbol, start=start, end=end)
         px = pd.DataFrame.rename(px, columns={'Adj Close': 'adj_close'})
         px.columns.name = symbol
         rets = px.adj_close.pct_change().dropna()
+        return rets
 
-        if symbol == 'SPY':
-            try:
-                rets.to_hdf(filepath, 'df')
-            except:
-                pass
+    if symbol == 'SPY':
+        filepath = data_path('spy.h5')
+        # Is cache recent enough?
+        if pd.to_datetime(getmtime(filepath), unit='s') >= end:
+            rets = pd.read_hdf(filepath, 'df')
+        else:
+            # Download most-recent SPY to update cache
+            rets = get_symbol_from_yahoo(symbol, start='1/1/1970',
+                                         end=datetime.now())
+            rets.to_hdf(filepath, 'df')
```

Comment:
I'm fairly certain that this will throw an exception in the case when the directory is write-protected. This was giving me some trouble before.